core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/jdbc/JDBCUserStoreManager.java

```diff
@@ -2784,52 +2787,62 @@ protected String preparePassword(String password, String saltValue) throws UserS
-            }
-
-            String digestFunction = realmConfig.getUserStoreProperties().get(JDBCRealmConstants.DIGEST_FUNCTION);
-            if (digestFunction != null) {
-                if (digestFunction.equals(UserCoreConstants.RealmConfig.PASSWORD_HASH_METHOD_PLAIN_TEXT)) {
-                    passwordString = new String(credentialObj.getChars());
-                    return passwordString;
+        Map<String, Object> userStoreProperties = Collections.unmodifiableMap(realmConfig.getUserStoreProperties());
+        String digestFunction = userStoreProperties.get(JDBCRealmConstants.DIGEST_FUNCTION).toString();
+        if (digestFunction != null) {
+            Map<String, HashProvider> hashProviderMap = UserStoreMgtDataHolder.getInstance().getHashProviderMap();
+            if (!hashProviderMap.containsKey(digestFunction)) {
+                try {
+                    if (digestFunction.equals(UserCoreConstants.RealmConfig.PASSWORD_HASH_METHOD_PLAIN_TEXT)) {
+                        passwordHash = new String(credentialObj.getChars());
+                    } else {
+                        if (saltValue != null) {
+                            credentialObj.addChars(saltValue.toCharArray());
+                        }
+                        MessageDigest digest = MessageDigest.getInstance(digestFunction);
+                        byte[] byteValue = digest.digest(credentialObj.getBytes());
+                        passwordHash = Base64.encode(byteValue);
+                    }
+                } catch (NoSuchAlgorithmException e) {
+                    String msg = "Error occurred while preparing password.";
+                    if (log.isDebugEnabled()) {
+                        log.debug(msg, e);
+                    }
+                    throw new UserStoreException(msg, e);
                 }
-
-                MessageDigest digest = MessageDigest.getInstance(digestFunction);
-                byte[] byteValue = digest.digest(credentialObj.getBytes());
-                passwordString = Base64.encode(byteValue);
             } else {
-                passwordString = new String(credentialObj.getChars());
-            }
-
-            return passwordString;
-        } catch (NoSuchAlgorithmException e) {
-            String msg = "Error occurred while preparing password.";
-            if (log.isDebugEnabled()) {
-                log.debug(msg, e);
+                String passwordString = String.valueOf(credentialObj.getChars());
+                HashProvider hashProvider = UserStoreMgtDataHolder.getInstance().getHashProvider(digestFunction);
+                try {
+                    byte[] hashByteArray = hashProvider.getHash(passwordString, saltValue, userStoreProperties);
+                    passwordHash = Base64.encode(hashByteArray);
+                } catch (HashProviderException e) {
+                    String msg = "Error in providing hashing functionality.";
```

Comment:
Shall we improve this as `Error occurred while preparing password`? I think we do not need another variable for this. WDYT?